### ISSUE TITLE
[data]'DataFrame' object has no attribute 'num_columns' using StandardScaler

### ISSUE THREAD

**Issue by wxy117** [ray-project/ray#39206]
### What happened + What you expected to happen

Hello,
I've ran into a pandas AttributeError when I use StandardScaler.

The logs are here.
 ```
File "/home/anaconda3/envs/ray/lib/python3.9/site-packages/ray/data/_internal/planner/exchange/aggregate_task_spec.py", line 71, in reduce
    return BlockAccessor.for_block(mapper_outputs[0]).aggregate_combined_blocks(
  File "/home/anaconda3/envs/ray/lib/python3.9/site-packages/ray/data/_internal/arrow_block.py", line 589, in aggregate_combined_blocks
    next_row = next(iter)
  File "/home/anaconda3/envs/ray/lib/python3.9/heapq.py", line 373, in merge
    value = next()
  File "/home/anaconda3/envs/ray/lib/python3.9/site-packages/ray/data/_internal/table_block.py", line 193, in __next__
    if self._cur < outer.num_rows():
  File "/home/anaconda3/envs/ray/lib/python3.9/site-packages/ray/data/_internal/arrow_block.py", line 282, in num_rows
    return self._table.num_rows if self._table.num_columns > 0 else 0
  File "/home/anaconda3/envs/ray/lib/python3.9/site-packages/pandas/core/generic.py", line 6202, in __getattr__
    return object.__getattribute__(self, name)
AttributeError: 'DataFrame' object has no attribute 'num_columns'
```

### Versions / Dependencies

ray==2.6.3

### Reproduction script

```
from ray.data.preprocessors import StandardScaler

def map_test(data):
    print(data)
    return data

ds = ray.data.from_items([
    {"A": x % 3, "B": x} for x in range(100)]).groupby(
    "A").map_groups(map_test).map_batches(map_test, batch_format='pandas')

ss = StandardScaler(['A', 'B'])
ss.fit_transform(ds)
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by PRESIDENT810** [comment]
I dig a little into this. I think it's because in this file `ray/data/_internal/planner/exchange/aggregate_task_spec.py`:
```
@staticmethod
    def reduce(
        key: Optional[str],
        aggs: List[AggregateFn],
        *mapper_outputs: List[Block],
        partial_reduce: bool = False,
    ) -> Tuple[Block, BlockMetadata]:
        return BlockAccessor.for_block(mapper_outputs[0]).aggregate_combined_blocks(
            list(mapper_outputs), key, aggs, finalize=not partial_reduce
        )
```
It uses the first item in mapper_outputs to get a BlockAccessor for either pandas or pyarrow tables, but in your case, mapper_outputs contains both `pandas.DataFrame` and `pyarraw.Table`, and trying to use ArrowBlockAccessor on a pandas.DataFrame caused this error, which is obvious because in your error logs it's clearly using methods in `arrow_block.py`.

**Comment by PRESIDENT810** [comment]
<img width="901" alt="image" src="https://github.com/ray-project/ray/assets/44538064/c8049f81-c242-4d92-a079-ce50b54c3ac4">

In the example you provided, after ds get materialized, its underlying blocks are mixed with both pyarrow.Table and pandas.DataFrame, and the function above only uses the first item do decide which BlockAccessor it should use, which is pyarrow.Table. Therefore when it handles a pandas.DataFrame an error is triggered.

I think from the APIs here, mixing pyarrow.Table and pandas.DataFrame in List[Block] should legal. Probably this API shouldn't assume all Blocks are the same format.

**Comment by wxy117** [comment]
> <img alt="image" width="901" src="https://user-images.githubusercontent.com/44538064/265942821-c8049f81-c242-4d92-a079-ce50b54c3ac4.png">
> In the example you provided, after ds get materialized, its underlying blocks are mixed with both pyarrow.Table and pandas.DataFrame, and the function above only uses the first item do decide which BlockAccessor it should use, which is pyarrow.Table. Therefore when it handles a pandas.DataFrame an error is triggered.
> 
> I think from the APIs here, mixing pyarrow.Table and pandas.DataFrame in List[Block] should legal. Probably this API shouldn't assume all Blocks are the same format. However if you want to get around this quickly, I think you can just drop `batch_format='pandas'` so all blocks will be pyarrow.Table and everything will be fine. You can always convert the dataset to pandas using `ds.to_pandas()`.

Thanks, it helps a lot.

**Comment by wxy117** [comment]
The error still persists in version 2.7.0, and the commit in #39580 has resolved it. I kindly request for someone to merge it into the master branch.

**Comment by PRESIDENT810** [comment]
> The error still persists in version 2.7.0, and the commit in #39580 has resolved it. I kindly request for someone to merge it into the master branch.

No, that commit is broken (causing some failure in other tests). I will try to fix it when I have some time to investigate into that...

**Comment by scottjlee** [comment]
The Ray Data team has also seen several other issues related to this one and have discussed how to best tackle all of them.

As you suggested above, the issue is that the blocks are of different types in the `reduce()` method of `aggregate_task_spec.py`. Before we call `aggregate_combined_blocks()` [here](https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/planner/exchange/aggregate_task_spec.py#L70-L72), if we convert all the input blocks to Arrow first, this should ensure all blocks are Arrow before being aggregated.

@PRESIDENT810 if you have further questions on this, please let me know here and I can provide further guidance

**Comment by michaelhly** [comment]
@scottjlee this issue can be fixed like this:
```
diff --git a/python/ray/data/_internal/arrow_block.py b/python/ray/data/_internal/arrow_block.py
index 2078aea575..6380068ff0 100644
--- a/python/ray/data/_internal/arrow_block.py
+++ b/python/ray/data/_internal/arrow_block.py
@@ -552,7 +552,7 @@ class ArrowBlockAccessor(TableBlockAccessor):
 
         iter = heapq.merge(
             *[
-                ArrowBlockAccessor(block).iter_rows(public_row_format=False)
+                BlockAccessor.for_arrow_block(block).iter_rows(public_row_format=False)
                 for block in blocks
             ],
             key=key_fn,
diff --git a/python/ray/data/block.py b/python/ray/data/block.py
index ea5937ace8..124f7e336b 100644
--- a/python/ray/data/block.py
+++ b/python/ray/data/block.py
@@ -373,6 +373,16 @@ class BlockAccessor:
                 return pd.DataFrame(dict(batch))
         return batch
 
+    @staticmethod
+    def for_arrow_block(block: Block) -> "ArrowBlockAccessor":
+        acc = BlockAccessor.for_block(block)
+        import pyarrow as pa
+
+        if not isinstance(block, pa.Table):
+            return BlockAccessor.for_block(acc.to_arrow())
+
+        return acc
+
     @staticmethod
     def for_block(block: Block) -> "BlockAccessor[T]":
         """Create a block accessor for the given block."""
```

We make sure all blocks inside the iterator are arrow blocks. What do you think?

**Comment by michaelhly** [comment]
But if `batch_format='pandas'` isn't specified. We don't get any errors ...

**Comment by PRESIDENT810** [comment]
@scottjlee I tried to fix this but I just realized #39960 seems to be a more generalized way to solve this issue, so probably we should wait for his approach to land.

Previously I tried to make BlocksToBatchesMapTransformFn.__call__() emit the empty block in user specified format in #39580. In the original code, it seems that when emitting the empty_block, the empty_block is not converted to `self._batch_format` but uses the format of the first block, whereas `formatted_batch_iter` gets converted to the specified format. I was trying to make this function always return blocks using the user specified format, which makes reduce() function doesn't get heterogeneous blocks as its inputs and can also solve this issue.
```
        empty_block = BlockAccessor.for_block(first).builder().build()
        # Don't hold the first block in memory, so we reset the reference.
        first = None

        # Ensure that zero-copy batch views are copied so mutating UDFs don't error.
        formatted_batch_iter = batch_blocks(
            blocks=blocks,
            stats=None,
            batch_size=self._batch_size,
            batch_format=self._batch_format,
            ensure_copy=self._ensure_copy,
        )

        first = next(formatted_batch_iter, None)
        if first is None:
            # If the input blocks are all empty, then yield an empty block with same
            # format as the input blocks.
            return [empty_block]
        else:
            return itertools.chain([first], formatted_batch_iter)
```
However, that cause a few test to fail, and I still don't quite understand why. If you got time, would you please give me some hint about where it could go wrong?

**Comment by anyscalesam** [comment]
Reviewed - for Ray 2.10 we should think of a holistic fix for all of these multi block type shuffle issues. cc @c21
